php/libraries/LorisForm.class.inc

```diff
@@ -1956,31 +1956,36 @@ class LorisForm
                 continue;
             }
 
-            if ($el['type'] === 'date'
-                && array_key_exists('options', $el)
-                && ((strtolower($el['options']['format']) == 'y'))
-            ) {
-                    $value = $this->getValue($name);
-                if ($value) {
-                    $arr[$name] = sprintf("%s-00-00", $value);
-                } else {
+            if ($el['type'] === 'date' && array_key_exists('options', $el)) {
+                $value = $this->getValue($name);
+                $date  = $value;
+
+                if (!$value) {
                     $arr[$name] = '';
-                }
                     continue;
-            }
+                }
 
-            if ($el['type'] === 'date'
-                && array_key_exists('options', $el)
-                && ((strtolower($el['options']['format']) == 'my')
-                || (strtolower($el['options']['format']) == 'ym'))
-            ) {
-                    $value = $this->getValue($name);
-                if ($value) {
-                    $arr[$name] = sprintf("%s-00", $value);
-                } else {
-                    $arr[$name] = '';
+                // Convert year to a valid SQL format (i.e yyyy-mm-dd)
+                if (strtolower($el['options']['format']) === 'y') {
+                    $date = sprintf("%s-01-01", $value);
                 }
-                    continue;
+                // Convert month-year to a valid SQL format (i.e yyyy-mm-dd)
+                if ((strtolower($el['options']['format']) === 'my')
+                    || (strtolower($el['options']['format']) === 'ym')
+                ) {
+                    $date = sprintf("%s-01", $value);
+                }
+                // Make sure the date is valid
+                $dt          = DateTime::createFromFormat('Y-m-d', $date);
+                $isValidDate = $dt && $dt->format('Y-m-d') == $date;
+
+                if (!$isValidDate) {
+                    throw new LorisException("Please input a valid date!");
```

Comment:
The problem is that the date validation was not done in `validate()` function which actually assigns the errors, but in a function that runs after error checking (which I updated for more robust checks. 

I'm open to take all the date related code and move it to validate, but I wanted @jstirling91 or @driusan to weight in as they are more familiar with the code.